File: src/main/java/com/eova/core/task/TaskIntercept.java
/**
 * Copyright (c) 2019 EOVA.CN. All rights reserved.
 * <p>
 * Licensed to: 卢中强(495993331@qq.com)
 * Licensed under the EPPL license: http://eova.cn/eppl.txt
 * Software copyright registration number:2018SR1012969
 * For authorization, please contact: admin@eova.cn
 */
package com.eova.core.task;

import com.eova.aop.AopContext;
import com.eova.aop.MetaObjectIntercept;
import com.eova.common.plugin.quartz.QuartzPlugin;
import com.eova.common.utils.xx;
import com.eova.model.Task;
import org.quartz.*;

public class TaskIntercept extends MetaObjectIntercept {

    @SuppressWarnings({"unchecked", "rawtypes"})
    @Override
    public String addAfter(AopContext ac) throws Exception {
        String className = ac.record.getStr("clazz");
        String exp = ac.record.getStr("exp");

        Class clazz = Class.forName(className);
        JobDetail job = JobBuilder.newJob(clazz).withIdentity(className, className).build();
        CronTrigger trigger = TriggerBuilder.newTrigger().withIdentity(className, className).withSchedule(CronScheduleBuilder.cronSchedule(exp)).build();

        // 新增任务
        QuartzPlugin.scheduler.scheduleJob(job, trigger);
        // 默认暂停
        QuartzPlugin.scheduler.pauseJob(job.getKey());

        return null;
    }

    @Override
    public String deleteAfter(AopContext ac) throws Exception {
        String className = ac.record.getStr("clazz");

        // 删除任务(自动删除关联触发器)
        JobKey jobKey = JobKey.jobKey(className, className);

        QuartzPlugin.scheduler.deleteJob(jobKey);

        return null;
    }

    @Override
    public String updateAfter(AopContext ac) throws Exception {
        String className = ac.record.getStr("clazz");
        String exp = ac.record.getStr("exp");

        // 表达式调度构建器
        CronScheduleBuilder scheduleBuilder = CronScheduleBuilder.cronSchedule(exp);

        // 重新构建触发器覆盖原触发器
        TriggerKey triggerKey = TriggerKey.triggerKey(className, className);
        CronTrigger trigger = (CronTrigger) QuartzPlugin.scheduler.getTrigger(triggerKey);
        trigger = trigger.getTriggerBuilder().withIdentity(triggerKey).withSchedule(scheduleBuilder).build();

        // 重置任务
        QuartzPlugin.scheduler.rescheduleJob(triggerKey, trigger);

        // 暂停
        QuartzPlugin.scheduler.pauseTrigger(triggerKey);

        // 状态变更为暂停
        Task.dao.updateState(xx.toInt(ac.record.get("id")), Task.STATE_STOP);

        return null;
    }

    @Override
    public String addBefore(AopContext ac) throws Exception {
        String cs = ac.record.getStr("clazz");
        boolean flag = Task.dao.isExist("select count(*) from eova_task where clazz = ?", cs);
        if (flag) {
            throw new Exception("Job实现类已经存在:" + cs);
        }

        return null;
    }

}

File: src/main/kotlin/cn/doit/natp/intercept/TaskIntercept.kt
@file:Suppress("UNCHECKED_CAST", "unused")

package cn.doit.natp.intercept

import cn.doit.natp.common.BaseMetaObjectIntercept
import com.eova.aop.AopContext
import com.eova.common.plugin.quartz.QuartzPlugin
import com.eova.common.utils.xx
import com.eova.model.Task
import org.quartz.*

class TaskIntercept : BaseMetaObjectIntercept() {

    override fun addAfter(ac: AopContext): String? {
        val className = ac.record.getStr("clazz")
        val exp = ac.record.getStr("exp")
        val clazz = Class.forName(className) as Class<out Job>
        val job = JobBuilder.newJob(clazz).withIdentity(className, className).build()
        val trigger = TriggerBuilder.newTrigger().withIdentity(className, className).withSchedule(CronScheduleBuilder.cronSchedule(exp)).build() as CronTrigger
        QuartzPlugin.scheduler.scheduleJob(job, trigger)
        QuartzPlugin.scheduler.pauseJob(job.key)

        return super.addAfter(ac)
    }

    override fun deleteAfter(ac: AopContext): String? {
        val className = ac.record.getStr("clazz")
        val jobKey = JobKey.jobKey(className, className)
        QuartzPlugin.scheduler.deleteJob(jobKey)
        return super.deleteAfter(ac)
    }

    @Throws(Exception::class)
    override fun updateAfter(ac: AopContext): String? {
        val className = ac.record.getStr("clazz")
        val exp = ac.record.getStr("exp")
        val scheduleBuilder = CronScheduleBuilder.cronSchedule(exp)
        val triggerKey = TriggerKey.triggerKey(className, className)
        var trigger = QuartzPlugin.scheduler.getTrigger(triggerKey) as CronTrigger
        trigger = trigger.triggerBuilder.withIdentity(triggerKey).withSchedule(scheduleBuilder).build() as CronTrigger
        QuartzPlugin.scheduler.rescheduleJob(triggerKey, trigger)
        QuartzPlugin.scheduler.pauseTrigger(triggerKey)
        Task.dao.updateState(xx.toInt(ac.record.get("id"))!!, 0)
        return super.updateAfter(ac)
    }

    @Throws(Exception::class)
    override fun addBefore(ac: AopContext): String? {
        val cs = ac.record.getStr("clazz")
        val flag = Task.dao.isExist("select count(*) from eova_task WHERE clazz = ?", cs)
        return if (flag) {
            throw Exception("Job实现类已经存在:$cs")
        } else {
            null
        }
    }

}